### ISSUE TITLE
```tf.distribute.MirroredStrategy``` is stuck.

### ISSUE THREAD

**Issue by DachuanZhao** [tensorflow/tensorflow#44976]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu20.04
- TensorFlow installed from (source or binary): pip install tensorflow
- TensorFlow version (use command below): 2.3.1
- Python version: 3.8.3
- CUDA/cuDNN version: 10.1
- GPU model and memory: gtx1070 * 2

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
```
== check python ===================================================
python version: 3.8.3
python branch:
python build version: ('default', 'Jul  2 2020 16:21:59')
python compiler version: GCC 7.3.0
python implementation: CPython


== check os platform ===============================================

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 7.5.0-6ubuntu2) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


== check pips ===================================================
numpy                              1.18.5
numpydoc                           1.1.0
protobuf                           3.13.0
tensorflow                         2.3.1
tensorflow-estimator               2.3.0
tensorflow-serving-api             2.3.0

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.version.VERSION = 2.3.1
tf.version.GIT_VERSION = v2.3.0-54-gfcc4b966f1
tf.version.COMPILER_VERSION = 7.3.1 20180303
    194729:     find library=libpthread.so.0 [0]; searching
    194729:      search path=/home/zhaodachuan/anaconda3/bin/../lib/tls/x86_64/x86_64:/home/zhaodachuan/anaconda3/bin/../lib/tls/x86_64:/home/zhaodachuan/anaconda3/bin/../lib/tls/x86_64:/home/zhaodachuan/anaconda3/bin/../lib/tls:/home/zhaodachuan/anaconda3/bin/../lib/x86_64/x86_64:/home/zhaodachuan/anaconda3/bin/../lib/x86_64:/home/zhaodachuan/anaconda3/bin/../lib/x86_64:/home/zhaodachuan/anaconda3/bin/../lib         (RPATH from file /home/zhaodachuan/anaconda3/bin/python)
    194729:       trying file=/home/zhaodachuan/anaconda3/bin/../lib/tls/x86_64/x86_64/libpthread.so.0
    194729:       trying file=/home/zhaodachuan/anaconda3/bin/../lib/tls/x86_64/libpthread.so.0
    194729:       trying file=/home/zhaodachuan/anaconda3/bin/../lib/tls/x86_64/libpthread.so.0
    194729:       trying file=/home/zhaodachuan/anaconda3/bin/../lib/tls/libpthread.so.0
== env ==========================================================
LD_LIBRARY_PATH /usr/local/cuda/lib64
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Wed Nov 18 19:10:46 2020
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.80.02    Driver Version: 450.80.02    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce GTX 1070    Off  | 00000000:23:00.0 Off |                  N/A |
| 42%   53C    P2    55W / 180W |   7709MiB /  8119MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 1070    Off  | 00000000:2D:00.0  On |                  N/A |
| 47%   65C    P2    60W / 180W |   7716MiB /  8111MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1041      G   /usr/lib/xorg/Xorg                  4MiB |
|    0   N/A  N/A    171379      G   /usr/lib/xorg/Xorg                  4MiB |
|    0   N/A  N/A    193747      C   ...huan/anaconda3/bin/python     7695MiB |
|    1   N/A  N/A      1041      G   /usr/lib/xorg/Xorg                 35MiB |
|    1   N/A  N/A    171379      G   /usr/lib/xorg/Xorg                 77MiB |
|    1   N/A  N/A    171513      G   /usr/bin/gnome-shell              119MiB |
|    1   N/A  N/A    177947      G   /usr/bin/nvidia-settings            3MiB |
|    1   N/A  N/A    193747      C   ...huan/anaconda3/bin/python     7461MiB |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudart_static.a
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudart.so.10.1.243
/usr/local/cuda-10.1/doc/man/man7/libcudart.so.7
/usr/local/cuda-10.1/doc/man/man7/libcudart.7

== tensorflow installed from info ==================
Name: tensorflow
Version: 2.3.1
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /home/zhaodachuan/anaconda3/lib/python3.8/site-packages
Required-by: tensorflow-serving-api

== python version  ==============================================
(major, minor, micro, releaselevel, serial)
(3, 8, 3, 'final', 0)

== bazel version  ===============================================
```
You can also obtain the TensorFlow version with:
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`
```
2020-11-18 19:09:31.423127: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
v2.3.0-54-gfcc4b966f1 2.3.1
```

**Describe the current behavior**
When I run the code 
```Python
import tensorflow as tf
gpus = tf.config.experimental.list_physical_devices('GPU')
tf.config.experimental.set_visible_devices(gpus, 'GPU')
print(gpus)
# output : 
# [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')]
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=500)
```
, it's stuck in 
```
Epoch 1/500
INFO:tensorflow:batch_all_reduce: 4 all-reduces with algorithm = nccl, num_packs = 1
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:batch_all_reduce: 4 all-reduces with algorithm = nccl, num_packs = 1
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
```
When I change to 
```Python
mirrored_strategy = tf.distribute.MirroredStrategy(cross_device_ops=tf.distribute.ReductionToOneDevice())
```
 , it can work but it runs slower than one gpu and the loss will become nan quickly
```
Epoch 1/500
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
  15/1875 [..............................] - ETA: 13:46 - loss: nan - accuracy: 0.1583

```
May the  same issue : 
[https://github.com/tensorflow/tensorflow/issues/37561](https://github.com/tensorflow/tensorflow/issues/37561)
[https://github.com/tensorflow/tensorflow/issues/38982](https://github.com/tensorflow/tensorflow/issues/38982)

**Comment by aaronmondal** [comment]
This is very likely an NCCL error, not a TensorFlow error. Try running the [Nvidia NCCL tests](https://github.com/NVIDIA/nccl-tests), e.g. the `all_reduce` test. If that gets stuck, you probably need to change your BIOS settings.

**Comment by DachuanZhao** [comment]
> This is very likely an NCCL error, not a TensorFlow error. Try running the [Nvidia NCCL tests](https://github.com/NVIDIA/nccl-tests), e.g. the `all_reduce` test. If that gets stuck, you probably need to change your BIOS settings.

Yes , it gets stuck ...... What should I do ? I use MSI AMD X570 motherboard

**Comment by aaronmondal** [comment]
From the [NCCL troubleshooting section](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/troubleshooting.html):

> IO virtualization (also known as, VT-d or IOMMU) can interfere with GPU Direct by redirecting all PCI point-to-point traffic to the CPU root complex, causing a significant performance reduction or even a hang. 

A possible quick fix may be simply IOMMU in your BIOS. Furthermore, if you run in a container:

> When using NCCL inside a container, it is recommended that you increase these resources by issuing: `–shm-size=1g –ulimit memlock=-1` in the command line to `nvidia-docker run`.

**Comment by DachuanZhao** [comment]
> IO virtualization

I turn off IOMMU then it works ! thanks
